public/app/plugins/panel/datagrid/DataGridPanel.tsx

```diff
@@ -0,0 +1,456 @@
+  HeaderClickedEventArgs,
+  SizedGridColumn,
+} from '@glideapps/glide-data-grid';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import {
+  ArrayVector,
+  Field,
+  MutableDataFrame,
+  PanelProps,
+  GrafanaTheme2,
+  getFieldDisplayName,
+  FieldType,
+} from '@grafana/data';
+import { PanelDataErrorView } from '@grafana/runtime';
+// eslint-disable-next-line import/order
+import { useTheme2 } from '@grafana/ui';
+
+import '@glideapps/glide-data-grid/dist/index.css';
+
+import { AddColumn } from './components/AddColumn';
+import { DatagridContextMenu } from './components/DatagridContextMenu';
+import { RenameColumnCell } from './components/RenameColumnCell';
+import { PanelOptions } from './panelcfg.gen';
+import {
+  RenameColumnInputData,
+  clearCellsFromRangeSelection,
+  DatagridContextMenuData,
+  deleteRows,
+  EMPTY_CELL,
+  EMPTY_GRID_SELECTION,
+  getCellWidth,
+  isDatagridEditEnabled,
+  publishSnapshot,
+  RIGHT_ELEMENT_PROPS,
+  TRAILING_ROW_OPTIONS,
+} from './utils';
+
+interface Props extends PanelProps<PanelOptions> {}
+
+export const DataGridPanel: React.FC<Props> = ({ options, data, id, fieldConfig }) => {
+  const [columnWidths, setColumnWidths] = useState<Map<number, number>>(new Map());
+  const [columns, setColumns] = useState<SizedGridColumn[]>([]);
+  const [contextMenuData, setContextMenuData] = useState<DatagridContextMenuData>({ isContextMenuOpen: false });
+  const [renameColumnInputData, setRenameColumnInputData] = useState<RenameColumnInputData>({ isInputOpen: false });
+  const [gridSelection, setGridSelection] = useState<GridSelection>(EMPTY_GRID_SELECTION);
+  const [columnFreezeIndex, setColumnFreezeIndex] = useState<number>(0);
+  const [toggleSearch, setToggleSearch] = useState<boolean>(false);
+  const [isResizeInProgress, setIsResizeInProgress] = useState<boolean>(false);
+
+  const frame = data.series[options.selectedSeries ?? 0];
+
+  const theme = useTheme2();
+  const gridTheme = {
+    accentColor: theme.colors.primary.main,
+    accentFg: theme.colors.secondary.main,
+    textDark: theme.colors.text.primary,
+    textMedium: theme.colors.text.primary,
+    textLight: theme.colors.text.primary,
+    textBubble: theme.colors.text.primary,
+    textHeader: theme.colors.text.primary,
+    bgCell: theme.colors.background.primary,
+    bgCellMedium: theme.colors.background.primary,
+    bgHeader: theme.colors.background.secondary,
+    bgHeaderHasFocus: theme.colors.background.secondary,
+    bgHeaderHovered: theme.colors.background.secondary,
+  };
+
+  const setGridColumns = useCallback(() => {
+    const typeToIconMap: Map<string, GridColumnIcon> = new Map([
+      [FieldType.number, GridColumnIcon.HeaderNumber],
+      [FieldType.string, GridColumnIcon.HeaderTextTemplate],
+      [FieldType.boolean, GridColumnIcon.HeaderBoolean],
+      [FieldType.time, GridColumnIcon.HeaderDate],
+      [FieldType.other, GridColumnIcon.HeaderReference],
+    ]);
+
+    if (!frame) {
```

Comment:
I would move this early return to before the `typeToIconMap` declaration, that way, we don't define something that possibly won't be used anyway.